packages/opentelemetry-basic-tracer/src/export/BatchSpanProcessor.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TraceOptions } from '@opentelemetry/types';
+import { SpanProcessor } from '../SpanProcessor';
+import { SpanExporter } from './SpanExporter';
+import { Span } from '../Span';
+import { ReadableSpan } from './ReadableSpan';
+import { unrefTimer } from '../platform';
+import { BufferConfig } from '../types';
+
+const DEFAULT_BUFFER_SIZE = 100;
+const DEFAULT_BUFFER_TIMEOUT = 20000;
+
+/**
+ * Implementation of the {@link SpanProcessor} that batches spans exported by
+ * the SDK then pushes them to the exporter pipeline.
+ */
+export class BatchSpanProcessor implements SpanProcessor {
+  private readonly _bufferSize: number = DEFAULT_BUFFER_SIZE;
+  private readonly _bufferTimeout: number = DEFAULT_BUFFER_TIMEOUT;
+  private _finishedSpans: ReadableSpan[] = [];
+  private _resetTimeout = false;
+  private _bufferTimeoutInProgress = false;
+
+  constructor(private readonly _exporter: SpanExporter, config?: BufferConfig) {
+    if (config) {
+      this._bufferSize = isNaN(Number(config.bufferSize))
```

Comment:
why the hassle or casting? why not accepting only `typeof === 'number'`?

General question: Can we assert on it? Like, `assert.optionalNumber(config.bufferSize)`? I know assertation and tracers are not friends, but it can be an operational concern falling back to defaults. Are startup time assertations okay? And only runtime not?

Reply:
Not sure of the perf impact of this, only consideration I can think of where it could accrue pain would be serverless/CLIs/other-short-life-things. WDYT?